### PR TITLE
File in zip source

### PR DESCRIPTION
**Pull request by rgabbard**

### CODE REVIEW

**Comment by rgabbard** [comment]
@berquist : This extends the previous MR for convenience.  Review the first one and I'll merge it, and then review this one.

File: vistautils/io_utils.py [hunk 1]
```diff
@@ -148,6 +171,15 @@ def is_empty(self) -> bool:
         return os.path.getsize(self._path) == 0
 
 
+@attrs(slots=True, frozen=True, auto_attribs=True)
+class _FromByteSourceCharSource(CharSource):
```
Comment (berquist) [inline vistautils/io_utils.py]: This was confusing at first until I realized it's a `CharSource` from a `ByteSource`. How about
```suggestion
class _CharSourceFromByteSource(CharSource):
```

File: vistautils/io_utils.py [hunk 2]
```diff
@@ -296,6 +342,37 @@ def is_empty(self) -> bool:
         return os.path.getsize(self._path) == 0
 
 
+@attrs(slots=True, frozen=True)
+class _FileInZipFilePathByteSource(ByteSource):
```
Comment (berquist) [inline vistautils/io_utils.py]: Same thing here.

File: vistautils/io_utils.py [hunk 3]
```diff
@@ -296,6 +342,37 @@ def is_empty(self) -> bool:
...
+        def new_close(self):
+            old_close()
+            zip_file.close()
+
+        ret.close = types.MethodType(new_close, ret)  # type: ignore
+        return ret  # type: ignore
+
+
+@attrs(slots=True, frozen=True)
+class _FileInOpenZipFileByteSource(ByteSource):
```
Comment (berquist) [inline vistautils/io_utils.py]: Same thing here.

File: vistautils/io_utils.py [hunk 4]
```diff
@@ -296,6 +342,37 @@ def is_empty(self) -> bool:
...
 
+@attrs(slots=True, frozen=True)
+class _FileInZipFilePathByteSource(ByteSource):
+    _zip_path = attrib_instance_of(Path)
+    _path_within_zip = attrib_instance_of(str)
+
+    def open(self) -> BytesIO:
+        # pylint:disable=not-callable
+        # pylint:disable=unused-argument
+        zip_file = ZipFile(self._zip_path, 'a')
```
Comment (berquist) [inline vistautils/io_utils.py]: I assume you don't want to write to the zip file:
```suggestion
        zip_file = ZipFile(self._zip_path, 'r')
```

**Pull request merged**